### PR TITLE
add decoders

### LINKED ISSUE #241: Implement Codec conversion
With client library

### PR DESCRIPTION
**Pull request by dmitry-s**
fixes #241

### CODE REVIEW

File: cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBinder.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner.r2dbc.v2;
+
+import com.google.cloud.spanner.Statement.Builder;
+
+public interface ClientLibraryBinder {
```
Comment (elefeint) [inline cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBinder.java:21]: `ClientLibraryTypeBinder`?

File: cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBinders.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import com.google.cloud.ByteArray;
+import com.google.cloud.Date;
+import com.google.cloud.Timestamp;
+import com.google.cloud.spanner.Statement;
+import com.google.cloud.spanner.r2dbc.statement.TypedNull;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+class ClientLibraryBinders {
```
Comment (elefeint) [inline cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBinders.java:12]: `ClientLibraryBinder`?

File: cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryValueBinder.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+  @Override
+  public boolean canEncode(Class type) {
+    Assert.requireNonNull(type, "type to encode must not be null");
+
+    return this.type.isAssignableFrom(type);
+  }
+
+  @Override
+  public void bind(Builder builder, String name, Object value) {
```
Comment (elefeint) [inline cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryValueBinder.java:27]: Fully qualify what type the `Builder` belongs to.

File: cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBinders.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        new ClientLibraryValueBinder(Boolean.class, (binder, val) -> binder.to((Boolean) val)));
+    binders.add(
+        new ClientLibraryValueBinder(ByteArray.class, (binder, val) -> binder.to((ByteArray) val)));
+    binders.add(new ClientLibraryValueBinder(Date.class, (binder, val) -> binder.to((Date) val)));
+    binders.add(
+        new ClientLibraryValueBinder(String.class, (binder, val) -> binder.to((String) val)));
+    binders.add(
+        new ClientLibraryValueBinder(Timestamp.class, (binder, val) -> binder.to((Timestamp) val)));
+
+    // There is technically one more supported type -  binder.to(Type type, @Nullable Struct value),
```
Comment (elefeint) [inline cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBinders.java:30]: `Struct` is an internal implementation detail, so R2DBC won't know about it. It's safe to disregard until we have a requirement for supporting custom object binding.

File: cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBinders.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    // but it is not clear how r2dbc could pass both the type and the value
+
+    return binders;
+  }
+
+  static void bind(Statement.Builder builder, String name, Object value) {
+    Class<?> valueClass = isTypedNull(value) ? ((TypedNull) value).getType() : value.getClass();
+
+    Optional<ClientLibraryBinder> encoderOptional =
```
Comment (elefeint) [inline cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBinders.java:39]: s/encoder/binder

File: cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBinders.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        binders.stream().filter(e -> e.canEncode(valueClass)).findFirst();
+    if (!encoderOptional.isPresent()) {
+      throw new RuntimeException("Can't find an encoder for type: " + valueClass);
```
Comment (elefeint) [inline cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBinders.java:42]: Extend `R2dbcNonTransientException` for a `BindingFailureException` or something,

File: cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryDecoder.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.Map;
+import java.util.function.BiFunction;
+
+public class ClientLibraryDecoder {
+  private static Map<Type, BiFunction<Struct, Integer, Object>> codecsMap = createCodecsMap();
+
+  private static Map<Type, BiFunction<Struct, Integer, Object>> arrayCodecsMap =
+      createArrayCodecsMap();
+
+  private static Map<Type, BiFunction<Struct, Integer, Object>> createArrayCodecsMap() {
```
Comment (elefeint) [inline cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryDecoder.java:16]: Call out in a comment that only 3 primitive array types are supported.

File: cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/SpannerClientLibraryRow.java [hunk 8]
```diff
@@ -1,24 +1,26 @@
...
 
 import com.google.cloud.spanner.Struct;
+import com.google.cloud.spanner.r2dbc.codecs.Codecs;
+import com.google.cloud.spanner.r2dbc.codecs.DefaultCodecs;
 import io.r2dbc.spi.Row;
 
 public class SpannerClientLibraryRow implements Row {
   private Struct rowFields;
 
+  private static final Codecs codecs = new DefaultCodecs();
```
Comment (elefeint) [inline cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/SpannerClientLibraryRow.java:11]: remove unused field

File: cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/ClientLibraryBasedIntegrationTest.java [hunk 9]
```diff
@@ -45,76 +43,89 @@ public static void setupSpannerTable() throws InterruptedException, ExecutionExc
...
-    Connection conn = Mono.from(connectionFactory.create())
-        .block();
+    Connection conn = Mono.from(connectionFactory.create()).block();
 
     StepVerifier.create(
-        Mono.from(conn.createStatement("SELECT count(*) FROM BOOKS").execute())
-          .flatMapMany(rs -> rs.map((row, rmeta) -> (Long)row.get(1)))
-    ).expectNext(Long.valueOf(0))
+            Mono.from(conn.createStatement("SELECT count(*) as count FROM BOOKS").execute())
+                .flatMapMany(rs -> rs.map((row, rmeta) -> (Long) row.get(1))))
```
Comment (elefeint) [inline cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/ClientLibraryBasedIntegrationTest.java:86]: Use overloaded `.get()` with type, remove cast.

File: cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBindersTest.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  /** Validates that every supported type binds successfully. */
+  @ParameterizedTest
+  @MethodSource("data")
+  <T> void binderTest(Class<T> type, Object value, BiConsumer<ValueBinder, Object> to) {
+    ClientLibraryBinders.bind(builder, "a", value);
+    ClientLibraryBinders.bind(builder, "b", new TypedNull(type));
+
+    to.accept(Mockito.verify(valueBinder, times(1)), value);
```
Comment (elefeint) [inline cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBindersTest.java:72]: Extract result of `Mockito.verify()` into a local variable named `instrumentedMethodCall` or something meaningful (although perhaps not as verbose as me).

File: cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBindersTest.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        arguments(
+            Timestamp.class,
+            Timestamp.ofTimeMicroseconds(123456),
+            (BiConsumer<ValueBinder, Object>) (binder, val) -> binder.to((Timestamp) val)));
+  }
+
+  /** Validates that every supported type binds successfully. */
+  @ParameterizedTest
+  @MethodSource("data")
+  <T> void binderTest(Class<T> type, Object value, BiConsumer<ValueBinder, Object> to) {
```
Comment (elefeint) [inline cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBindersTest.java:68]: Rename `to` to `verificationFunction` or similar.

File: cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBinders.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+import com.google.cloud.spanner.Statement;
+import com.google.cloud.spanner.r2dbc.statement.TypedNull;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+class ClientLibraryBinders {
+  private static List<ClientLibraryBinder> binders = buildEncoders();
+
+  private static List<ClientLibraryBinder> buildEncoders() {
```
[the remark this answers is not in the document]
Comment (dmitry-s) [reply cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryBinders.java:15]: I think it's better when it is close to the only method call - you don't have to jump around to see what is assigned to the `binders` field.

File: cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryDecoder.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+  /**
+   * Decodes result given index and type.
+   *
+   * @param struct the result struct
+   * @param index the index of the result to decode
+   * @param type the type of the result
+   * @param <T> the type of the result
+   * @return decoded value
+   */
+  public static <T> T decode(Struct struct, int index, Class<T> type) {
```
[the remark this answers is not in the document]
Comment (dmitry-s) [reply cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/ClientLibraryDecoder.java:59]: We are not in business of converting values to custom types. We rely on struct to provide the type of the result. The only choice that exists for user is for array results of `int64`, `float64` and `bool` - they can be decoded as primitive arrays or lists.

**Pull request merged**